[DAG] Handle truncated splat in isBoolConstant

davemgreen opened the pull request:
This allows truncated splat / buildvector in isBoolConstant, to allow certain not instructions to be recognized post-legalization, and allow vselect to optimize.

An override for x86 avx512 predicated vectors is required to avoid an infinite recursion from the code that detects zero vectors. From:
```
  // Check if the first operand is all zeros and Cond type is vXi1.
  // If this an avx512 target we can improve the use of zero masking by
  // swapping the operands and inverting the condition.
```

RKSimon commented on llvm/include/llvm/CodeGen/TargetLowering.h:
```diff
@@ -4375,6 +4375,8 @@ class LLVM_ABI TargetLowering : public TargetLoweringBase {
            Op.getOpcode() == ISD::SPLAT_VECTOR_PARTS;
   }
 
+  virtual bool isTargetCanonicalSelect(SDNode *N) const { return false; }
```
(style) add a description of what a canonical select means

arsenm replied:
This seems like a very specific hack, can you do this another way

phoebewang replied:
X86DAGToDAGISel sounds good to me. We have hundreds of AVX512 instructions, adding patterns for them is too verbose.

RKSimon replied:
I have no strong preference tbh and if you want to go with isTargetCanonicalSelect I won't object.

I agree with @arsenm that we add extra TLI hooks too quickly (and seldom get rid of ones that are no longer necessary), but #145724 is not ideal as an alternative.

davemgreen commented:
Hi - I'm Ok either way, up to you what you think is best. If you want to go with #145724 then I can give this a rebase. If you prefer not to support the negations in PreprocessISelDAG then we can go with this method. Either way thanks for the assistance.

davemgreen committed:
[DAG] Handle truncated splat in isBoolConstant

This allows truncated splat / buildvector in isBoolConstant, to allow certain
not instructions to be recognized post-legalization, and allow vselect to
optimize.

An override for x86 avx512 predicated vectors is required to avoid an infinite
recursion from the code that detects zero vectors. From:
```
  // Check if the first operand is all zeros and Cond type is vXi1.
  // If this an avx512 target we can improve the use of zero masking by
  // swapping the operands and inverting the condition.
```

davemgreen commented:
Thanks - I rebased and updated the sve tests that have changed in the meantime.

RKSimon reviewed (approved):
LGTM
